go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,285 @@
+	}
+	var token *Token
+	if err := json.Unmarshal(body, &token); err != nil {
+		return nil, fmt.Errorf("Failed to parse json in verify flow.")
+	}
+
+	return token, nil
+}
+
+// Helper struct used in sign JWT
+type sha256Opts struct{}
+
+func (r sha256Opts) HashFunc() crypto.Hash {
+	return crypto.SHA256
+}
+
+// Base 64 encode a block. The output doesn't contain the trailing double equal
+// signs.
+func base64Encode(b []byte) string {
+	return strings.TrimSuffix(base64.URLEncoding.EncodeToString(b), "==")
+}
+
+// Signer interface to support both rsa and ecdsa sign.
+type pkeyInterface interface {
+	Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) ([]byte, error)
+}
+
+// Creates a signed JWT token. Used for two-legged oauth token request.
+func createJWT(secret map[string]interface{}, scopes string, pkey pkeyInterface) (string, error) {
+	// A valid JWT has an "iat" timestamp and an "exp" timestamp. Get the current
+	// time to create these timestamps.
+	now := int(time.Now().Unix())
+
+	// Construct the JWT header, which contains the private key id in the service
+	// account secret.
+	header := map[string]string{
+		"typ": "JWT",
+		"alg": "RS256",
+		"kid": secret["private_key_id"].(string),
+	}
+
+	// Construct the JWT payload.
+	payload := map[string]string{
+		"aud":   secret["token_uri"].(string),
+		"scope": scopes,
+		"iat":   strconv.Itoa(now),
+		"exp":   strconv.Itoa(now + 3600),
+		"iss":   secret["client_email"].(string),
+	}
+
+	// Convert the header and payload map to json.
+	headerJson, err := json.Marshal(header)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert JWT header to json.")
+	}
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert JWT payload to json.")
+	}
+
+	// Convert the header and payload json to base64
+	headerB64 := base64Encode(headerJson)
+	payloadB64 := base64Encode(payloadJson)
+
+	// The first two segments of the JWT are signed. The signature is the third
+	// segment.
+	segments := headerB64 + "." + payloadB64
+
+	// sign the hash, instead of the actual segments.
+	hashed := sha256.Sum256([]byte(segments))
+	signedBytes, err := pkey.Sign(rand.Reader, hashed[:], crypto.SignerOpts(sha256Opts{}))
+	if err != nil {
+		return "", fmt.Errorf("Failed to sign JWT: %s", err)
+	}
+
+	// Generate the final JWT as
+	// base64(header) + "." + base64(payload) + "." + base64(signature)
+	return segments + "." + base64Encode(signedBytes), nil
+}
+
+// Interface for OAuth2 Client
+type Client interface {
+	// Get an OAuth 2 access token for the specified OAuth scopes.
+	// scopes: A space separated scope codes per OAuth 2.0 spec
+	// (https://tools.ietf.org/html/rfc6749).
+	// returns: the Token object. It does not contain scope information.
+	// GetToken returns a token or an error.
+	// GetToken must be safe for concurrent use by multiple goroutines.
+	// The returned Token must not be modified.
+	GetToken(scopes string) (*Token, error)
+}
```

Comment:
Suggest we copy TokenSource interface from /x/oauth2 to token.go. It will be easier for other people to port their code.